File: brain-move-web/src/games/musical-memory.ts
import { BaseGame } from './base-game'
import { Difficulty } from '../engine/difficulty'
import { Scoring } from '../engine/scoring'

export class MusicalMemory extends BaseGame {
  phase: "instruction" | "showing" | "input" | "result" = "instruction"
  sequence: number[] = []
  userInput: number[] = []
  inputIndex = 0
  showIndex = 0
  showing = false
  showTimer = 0
  flashTimer = 0
  flashNote = -1
  round = 0
  maxRounds = 10
  correctRounds = 0
  instructionTimer = 0
  private audioCtx: AudioContext | null = null
  colors = ["#e63c3c", "#3c82e6", "#3cc83c", "#e6d232"]
  freqs = [262, 330, 392, 523]
  labels = ["Do", "Mi", "Sol", "Do'"]

  constructor(difficulty: Difficulty, scoring: Scoring) {
    super(difficulty, scoring)
    this.name = "musical_memory"
    this.displayName = "Musical Memory"
  }

  private _getAudioCtx(): AudioContext {
    if (!this.audioCtx) this.audioCtx = new AudioContext()
    return this.audioCtx
  }

  private _playNote(noteIdx: number): void {
    try {
      const ctx = this._getAudioCtx()
      const osc = ctx.createOscillator()
      const gain = ctx.createGain()
      osc.connect(gain)
      gain.connect(ctx.destination)
      osc.frequency.value = this.freqs[noteIdx]
      osc.type = "sine"
      gain.gain.setValueAtTime(0.25, ctx.currentTime)
      gain.gain.exponentialRampToValueAtTime(0.01, ctx.currentTime + 0.35)
      osc.start(ctx.currentTime)
      osc.stop(ctx.currentTime + 0.35)
    } catch { /* audio not available */ }
  }

  start(): void {
    super.start()
    this.phase = "instruction"
    this.sequence = []
    this.round = 0
    this.correctRounds = 0
    this.inputIndex = 0
    this.showing = false
    this.instructionTimer = Date.now()
    this.feedbackMessage = "Ripeti la sequenza musicale!"
  }

  update(_poseData?: any, dt = 0): void {
    if (!this.running) return
    if (this.phase === "instruction") {
      if ((Date.now() - this.instructionTimer) / 1000 > 2) this._nextNote()
      return
    }
    if (this.showing) {
      this.showTimer -= dt
      this.flashTimer -= dt
      if (this.flashTimer <= 0) this.flashNote = -1
      if (this.showTimer <= 0) {
        this.showing = false
        this.phase = "input"
        this.inputIndex = 0
        this.feedbackMessage = `Ripeti! (${this.round}/${this.maxRounds})`
      }
    } else if (this.phase === "result") {
      this.showTimer -= dt
      if (this.showTimer <= 0) this._nextNote()
    }
  }

  private _nextNote(): void {
    this.round++
    this.sequence.push(Math.floor(Math.random() * 4))
    this.showing = true
    this.showIndex = 0
    this.showTimer = this.sequence.length * 0.8
    this._flashNext()
  }

  private _flashNext(): void {
    if (this.showIndex < this.sequence.length) {
      const note = this.sequence[this.showIndex]
      this.flashNote = note
      this.flashTimer = 0.6
      this._playNote(note)
      this.showIndex++
      setTimeout(() => this._flashNext(), 800)
    }
  }

  pressButton(noteIdx: number): void {
    if (this.phase !== "input") return
    this._playNote(noteIdx)
    if (noteIdx === this.sequence[this.inputIndex]) {
      this.inputIndex++
      if (this.inputIndex >= this.sequence.length) {
        this.correctRounds++
        this.feedbackMessage = `Corretto! ${this.correctRounds}/${this.round}`
        if (this.round >= this.maxRounds) this._finishGame()
        else { this.phase = "result"; this.showTimer = 1.5 }
      }
    } else {
      this.feedbackMessage = "Sbagliato!"
      this._finishGame()
    }
  }

  private _finishGame(): void {
    this.stop()
    const accuracy = this.correctRounds / Math.max(this.round, 1)
    this.result = { completed: true, accuracy, correctRounds: this.correctRounds, totalRounds: this.round, sequenceLength: this.sequence.length }
    this.difficulty.adapt(accuracy)
    this.scoring.addScore(this.name, Math.round(accuracy * 30))
  }

  getProgress(): number { return this.round / Math.max(this.maxRounds, 1) }
  getFlashNote(): number { return this.flashNote }
  getPhase(): string { return this.showing ? "showing" : this.phase }
}

File: brain-move/games/musical_memory.py
import random
import time
import math
import struct
from games.base_game import BaseGame


class MusicalMemory(BaseGame):
    COLORS = [
        (230, 60, 60),
        (60, 130, 230),
        (60, 200, 60),
        (230, 210, 50),
    ]
    FREQS = [262, 330, 392, 523]
    LABELS = ["Do", "Mi", "Sol", "Do'"]

    def __init__(self, difficulty_engine, scoring_engine):
        super().__init__(difficulty_engine, scoring_engine)
        self.name = "musical_memory"
        self.display_name = "Musical Memory"
        self.phase = "instruction"
        self.sequence = []
        self.user_input = []
        self.input_index = 0
        self.show_index = 0
        self.showing = False
        self.show_timer = 0
        self.flash_timer = 0
        self.flash_note = -1
        self.round = 0
        self.max_rounds = 10
        self.correct_rounds = 0
        self.instruction_timer = 0
        self.sounds = []
        self._init_sounds()

    def _init_sounds(self):
        for freq in self.FREQS:
            sound = self._make_tone(freq, 0.35)
            self.sounds.append(sound)

    def _make_tone(self, freq, duration):
        try:
            import pygame
            sample_rate = 22050
            num = int(sample_rate * duration)
            samples = []
            for i in range(num):
                t = i / sample_rate
                v = math.sin(2 * math.pi * freq * t)
                env = 1.0
                fade = int(sample_rate * 0.05)
                if i > num - fade:
                    env = (num - i) / fade
                v *= env * 0.25 * 32767
                samples.append(int(v))
            data = struct.pack("<" + "h" * num, *samples)
            return pygame.mixer.Sound(buffer=data)
        except Exception:
            return None

    def start(self):
        super().start()
        self.phase = "instruction"
        self.sequence = []
        self.round = 0
        self.correct_rounds = 0
        self.input_index = 0
        self.showing = False
        self.instruction_timer = time.time()
        self.feedback_message = "Ripeti la sequenza musicale!"

    def update(self, pose_data=None, dt=0):
        if not self.running:
            return

        if self.phase == "instruction":
            if time.time() - self.instruction_timer > 2.0:
                self._next_note()
            return

        if self.showing:
            self.show_timer -= dt
            self.flash_timer -= dt
            if self.flash_timer <= 0:
                self.flash_note = -1
            if self.show_timer <= 0:
                self.showing = False
                self.phase = "input"
                self.input_index = 0
                self.feedback_message = f"Ripeti! ({self.round}/{self.max_rounds})"

    def _next_note(self):
        self.round += 1
        self.sequence.append(random.randint(0, 3))
        self.showing = True
        self.show_index = 0
        self.show_timer = len(self.sequence) * 0.8
        self._flash_next()

    def _flash_next(self):
        if self.show_index < len(self.sequence):
            note = self.sequence[self.show_index]
            self.flash_note = note
            self.flash_timer = 0.6
            self._play_note(note)
            self.show_index += 1

    def _play_note(self, note_idx):
        if 0 <= note_idx < len(self.sounds) and self.sounds[note_idx]:
            try:
                self.sounds[note_idx].play()
            except Exception:
                pass

    def press_button(self, note_idx):
        if self.phase != "input":
            return
        self._play_note(note_idx)
        if note_idx == self.sequence[self.input_index]:
            self.input_index += 1
            if self.input_index >= len(self.sequence):
                self.correct_rounds += 1
                self.feedback_message = f"Corretto! {self.correct_rounds}/{self.round}"
                if self.round >= self.max_rounds:
                    self._finish_game()
                else:
                    self.phase = "result"
                    self.show_timer = 1.5
        else:
            self.feedback_message = f"Sbagliato! Era la sequenza corretta"
            self._finish_game()

    def _finish_game(self):
        self.stop()
        accuracy = self.correct_rounds / max(self.round, 1)
        self.result = {
            "completed": True,
            "accuracy": accuracy,
            "correct_rounds": self.correct_rounds,
            "total_rounds": self.round,
            "sequence_length": len(self.sequence),
        }
        self.difficulty.adapt(accuracy)
        self.scoring.add_score(self.name, int(accuracy * 30))

    def get_progress(self):
        return self.round / max(self.max_rounds, 1)

    def get_flash_note(self):
        return self.flash_note
